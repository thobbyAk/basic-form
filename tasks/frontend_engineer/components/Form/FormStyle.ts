import styled from "styled-components";


export const FormContainer = styled.form`
  margin-top: 40px;
  width: 300px;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

export const InputForm = styled.input`
    border: 1px solid #545f7d;
    border-radius: 8px;
    display: inline-flex;
    align-items: center;
    margin-bottom: 30px;
    box-sizing: border-box;
    overflow: hidden;
    transition: 0.3s ease;
    position: relative;
    outline: none;
    width: 100%;
    height: 100%;
    color: ${({ theme }) => theme.colors.primary};
    padding: 12px 20px;
    &:focus-within {
        border: 2px solid #545f7d;
      }
    &:hover {
    border-color: #5c068c;
    }
    
    &::-webkit-input-placeholder {
        font-family: inherit;
        color: #545f7d;
        font-weight: 400;
        font-size: 14px;
    }
    
`;

export const InputSelect = styled.select`
    border: 1px solid #545f7d;
    border-radius: 8px;
    display: inline-flex;
    align-items: center;
    margin-bottom: 30px;
    box-sizing: border-box;
    overflow: hidden;
    transition: 0.3s ease;
    position: relative;
    color: ${({ theme }) => theme.colors.primary};
    outline: none;
    width: 100%;
    height: 100%;
    padding: 12px 20px;
    &:focus-within {
        border: 2px solid #545f7d;
      }
    &:hover {
    border-color: #5c068c;
    }
    
    &::-webkit-input-placeholder {
        font-family: inherit;
        color: #545f7d;
        font-weight: 400;
        font-size: 14px;
    }
    
`;
export const CheckboxField = styled.div`  
    margin-bottom: 30px;
    display: inline-flex;
    flex-direction: row;
    gap: 4;
    label {
        color: ${({ theme }) => theme.colors.primary};
    }
    input {
        appearance: none;
        -webkit-appearance: none;
        -moz-appearance: none;
        width: 20px;
        height: 20px;
        position: relative;
        border: 1px solid #545f7d; 
        border-radius: 4px;
        &:before {
            content: '';
            position: absolute;
            top: 0px;
            left: 2px;
            width: 8px;
            height: 8px;
            border: 4px solid #5c068c; 
            border-top: none;
            border-right:none;
            transform: rotate(-46deg);
            opacity: 0;
          }
        
          &:checked:before {
            opacity: 1; 
          }
        
    }

`;

export const ButtonContainer = styled.button`  
   
    box-sizing: border-box;
    display: inline-block;
    color: white;
    background-color: #5c068c;
    font-size: 16px;
    line-height: 1.5;
    font-weight: 500;
    padding: 10px 18px;
    border-radius: 8px;
    cursor: pointer;
    transition: 0.3s ease;
    border: none;
    outline: none;
    width: 100%;

`;