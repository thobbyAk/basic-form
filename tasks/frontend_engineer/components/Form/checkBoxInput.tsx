import React, { HTMLProps, forwardRef, useState } from "react";

const CheckBoxInput = forwardRef<HTMLInputElement, HTMLProps<HTMLInputElement>>(
	({ ...props }, ref) => {
		const [agree, setAgree] = useState(false);
		return (
			<input
				checked={agree}
				onChange={(e) => setAgree((e.target as HTMLInputElement).checked)}
				{...props}
				ref={ref}
			/>
		);
	}
);

export default CheckBoxInput;
