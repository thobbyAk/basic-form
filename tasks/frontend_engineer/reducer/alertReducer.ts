import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export interface AlertState {
  message: string ;
  showAlert: boolean
}

const initialState: AlertState = {
  message: '',
  showAlert:false
};

export const alertSlice = createSlice({
  name: 'alert',
  initialState,
  reducers: {
    setMessage: (state, action: PayloadAction<AlertState>) => {
      state.message = action.payload?.message;
      state.showAlert = action.payload?.showAlert;

    },
  },
});

export const { setMessage } = alertSlice.actions;

