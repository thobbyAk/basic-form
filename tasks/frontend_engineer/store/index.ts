import rootReducer from '@/reducer';
import { configureStore } from '@reduxjs/toolkit';

export const store = configureStore({
    reducer: rootReducer
});

export type AppDispatch = typeof store.dispatch
export default store;