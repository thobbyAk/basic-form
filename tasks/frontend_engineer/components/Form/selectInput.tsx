import React, { SelectHTMLAttributes, forwardRef, useState } from "react";
import { InputSelect } from "./FormStyle";

const SelectInput = forwardRef<
	HTMLSelectElement,
	SelectHTMLAttributes<HTMLSelectElement>
>(({ ...props }, ref) => {
	const [category, setCategory] = useState("");
	return (
		<InputSelect
			ref={ref}
			value={category}
			onChange={(e) => setCategory((e.target as HTMLSelectElement).value)}
		>
			<option value="">Select category</option>
			<option value="Shoes">Shoes</option>
			<option value="Heels">Heels</option>
			<option value="Sandals">Sandals</option>
			<option value="Handbags">Handbags</option>
		</InputSelect>
	);
});

export default SelectInput;
