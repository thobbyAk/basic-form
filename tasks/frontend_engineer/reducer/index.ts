import { combineReducers } from "@reduxjs/toolkit";
import { alertSlice } from "./alertReducer";

const rootReducer = combineReducers({
    alertState: alertSlice.reducer
})
export default rootReducer;

export type IRootState = ReturnType<typeof rootReducer>;