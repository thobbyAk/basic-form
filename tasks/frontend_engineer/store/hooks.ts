import { useDispatch,useSelector, TypedUseSelectorHook } from "react-redux";
import { AppDispatch } from ".";
import { IRootState } from "@/reducer";

export const useAppDispatch = () => {
    const dispatch = useDispatch<AppDispatch>();
  
    return dispatch;
  };

  export const useAppStateSelector: TypedUseSelectorHook<IRootState> =
  useSelector;