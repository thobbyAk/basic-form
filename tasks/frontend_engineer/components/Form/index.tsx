import React, { useRef, useState } from "react";
import Input from "./input";
import Button from "./button";
import CheckBoxInput from "./checkBoxInput";
import SelectInput from "./selectInput";
import { CheckboxField, FormContainer } from "./FormStyle";
import { useAppDispatch } from "@/store/hooks";
import { AlertState, setMessage } from "@/reducer/alertReducer";

const Form = () => {
	const emailRef = useRef<HTMLInputElement>(null);
	const checkboxRef = useRef<HTMLInputElement>(null);
	const categoryRef = useRef<HTMLSelectElement>(null);
	const dispatch = useAppDispatch();

	const handleSubmit = (e: React.FormEvent) => {
		e.preventDefault();
		const payload: AlertState = {
			message: "Form submitted successfully",
			showAlert: true,
		};
		dispatch(setMessage(payload));
		clearAlertState();
		resetForm();
	};

	const clearAlertState = () => {
		const payload: AlertState = {
			message: "",
			showAlert: false,
		};
		setTimeout(() => {
			dispatch(setMessage(payload));
		}, 5000);
	};

	function resetForm() {
		if (emailRef.current) emailRef.current.value = "";
		if (checkboxRef.current) checkboxRef.current.checked = false;
		if (categoryRef.current) categoryRef.current.value = "";
	}

	return (
		<FormContainer onSubmit={handleSubmit}>
			<Input ref={emailRef} type="email" placeholder="Email address" required />
			<SelectInput ref={categoryRef} />
			<CheckboxField>
				<CheckBoxInput id="terms" ref={checkboxRef} type="checkbox" required />
				<label htmlFor="terms">I agree to the terms and conditions</label>
			</CheckboxField>

			<Button type="submit">Submit</Button>
		</FormContainer>
	);
};

export default Form;
