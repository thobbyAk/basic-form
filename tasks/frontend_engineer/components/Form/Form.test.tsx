import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import Form from ".";
import { Provider } from "react-redux";
import store from "@/store";
import { setMessage, alertSlice, AlertState } from "../../reducer/alertReducer";

const previousState: AlertState = {
	message: "",
	showAlert: false,
};

describe("Form Component", () => {
	test("it should match the snapshot and render", async () => {
		const { container } = render(
			<Provider store={store}>
				<Form />
			</Provider>
		);
		expect(container).toMatchSnapshot();
		expect(container).toBeInTheDocument();
	});
	test("submits the form with valid input", async () => {
		const { getByPlaceholderText, getByLabelText, getByText } = render(
			<Provider store={store}>
				<Form />
			</Provider>
		);

		const emailInput = getByPlaceholderText("Email address");
		fireEvent.change(emailInput, { target: { value: "testemail@yahoo.com" } });

		const categoryInput = getByText("Select category");
		fireEvent.change(categoryInput, { target: { value: "Shoes" } });

		const agreeCheckbox = getByLabelText("I agree to the terms and conditions");
		fireEvent.click(agreeCheckbox);

		const submitButton = getByText("Submit");
		fireEvent.click(submitButton);

		await waitFor(() => {
			const payload: AlertState = {
				message: "Form submitted successfully",
				showAlert: true,
			};
			expect(alertSlice.reducer(previousState, setMessage(payload))).toEqual(
				payload
			);
		});
	});
	test("the form should not submit if email or checkbox is invalid", async () => {
		const payload: AlertState = {
			message: "",
			showAlert: false,
		};
		const { getByPlaceholderText, getByLabelText, getByText } = render(
			<Provider store={store}>
				<Form />
			</Provider>
		);

		const emailInput = getByPlaceholderText("Email address");
		fireEvent.change(emailInput, { target: { value: "testemail@yahoo.co" } });

		const categoryInput = getByText("Select category");
		fireEvent.change(categoryInput, { target: { value: "Shoes" } });

		const submitButton = getByText("Submit");
		fireEvent.click(submitButton);

		await waitFor(() => {
			expect(alertSlice.reducer(previousState, setMessage(payload))).toEqual(
				payload
			);
		});
		const agreeCheckbox = getByLabelText("I agree to the terms and conditions");
		fireEvent.click(agreeCheckbox);
		fireEvent.change(emailInput, { target: { value: "email.com" } });
		fireEvent.click(submitButton);
		await waitFor(() => {
			expect(alertSlice.reducer(previousState, setMessage(payload))).toEqual(
				payload
			);
		});
	});
});
