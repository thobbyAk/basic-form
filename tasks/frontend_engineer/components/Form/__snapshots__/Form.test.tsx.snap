// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component it should match the snapshot and render 1`] = `
<div>
  <form
    class="sc-6dcabf1-0 eBDBiU"
  >
    <input
      class="sc-6dcabf1-1 gRPbSx"
      placeholder="Email address"
      required=""
      type="email"
      value=""
    />
    <select
      class="sc-6dcabf1-2 ethpxy"
    >
      <option
        value=""
      >
        Select category
      </option>
      <option
        value="Shoes"
      >
        Shoes
      </option>
      <option
        value="Heels"
      >
        Heels
      </option>
      <option
        value="Sandals"
      >
        Sandals
      </option>
      <option
        value="Handbags"
      >
        Handbags
      </option>
    </select>
    <div
      class="sc-6dcabf1-3 dGyqpl"
    >
      <input
        id="terms"
        required=""
        type="checkbox"
      />
      <label
        for="terms"
      >
        I agree to the terms and conditions
      </label>
    </div>
    <button
      class="sc-6dcabf1-4 izWVn"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
